lib/spack/spack/util/archive.py

```diff
@@ -238,3 +241,42 @@ def reproducible_tarfile_from_prefix(
                 add_file(tar, file_info, entry.path)
 
         dir_stack.extend(reversed(new_dirs))  # we pop, so reverse to stay alphabetical
+
+
+def _git_prefix(archive_path):
+    # This is an annoying method, but since we always have a prefix and can't gaurantee what
+    # it is we need this.
+    tar = which("tar", required=True)
+    paths = tar("-tf", archive_path, output=str, error=str, fail_on_error=False)
+    if paths:
+        paths = paths.strip().split()
+        for p in paths:
+            if p.endswith(".git/"):
+                return p[:-5]
+    return ""
+
+
+def retrieve_commit_from_archive(archive_path, ref):
+    """extract git data from an archive with out expanding it"""
+    assert os.path.isfile(archive_path)
+
+    commit = None
+    # try branch, tags then detached states
+    ref_paths = [f"refs/heads/{ref}/", f"refs/tags/{ref}/", "HEAD"]
+    tar = which("tar", required=True)
+    error_msg = f"Archive {archive_path} does not appear to contain git data"
+
+    prefix = _git_prefix(archive_path)
+
+    for ref_path in ref_paths:
+        try:
+            commit = tar(
+                "-Oxzf", archive_path, f"{prefix}.git/{ref_path}", output=str, error=str
+            ).strip()
+            if commit and len(commit) == 40:
+                break
+        except ProcessError:
+            pass
+    if not commit:
+        tty.warn(error_msg)
```

Comment:
```suggestion
    tar = which("tar", required=True)
    prefix = _git_prefix(archive_path)

    # try branch, tags then detached states
    for ref_path in [f"refs/heads/{ref}/", f"refs/tags/{ref}/", "HEAD"]:
        try:
            commit = tar(
                "-Oxzf", archive_path, f"{prefix}.git/{ref_path}", output=str, error=str
            ).strip()
            if commit and len(commit) == 40:
                break
        except ProcessError:
            pass
    if not commit:
        tty.warn(f"Archive {archive_path} does not appear to contain git data")
```
Suggestion for better legibility, let me know if you disagree and we'll chock it up to aesthetics.